### PR TITLE
adding graceful shutdown to the async event dispatcher

### PR DESCRIPTION
**Pull request by mdodsworth**
@vraja2 @aliabbasrizvi 
As we discussed. Enjoy!

Event dispatch actions are now wrapped in a `Runnable` and passed to the `ExecutorService` to enqueue and process.

The `shutdownAndAwaitTermination` method can the delegate through to the executor to do the necessary interrupting, cancelling, awaiting, etc.

### CODE REVIEW

File: core-api/src/test/java/com/optimizely/ab/bucketing/BucketerTest.java [hunk 1]
```diff
@@ -446,7 +446,7 @@ public void bucketUserNotInOverlappingGroupExperiment() throws Exception {
     }
 
     /**
-     * Verify that {@link Bucketer#bucket(Experiment,String} saves a variation of an experiment for a user
+     * Verify that {@link Bucketer#bucket(Experiment,String)} saves a variation of an experiment for a user
```
Comment (mdodsworth) [inline core-api/src/test/java/com/optimizely/ab/bucketing/BucketerTest.java]: some javadoc fixes spotted on the way through.

File: core-httpclient-impl/build.gradle [hunk 2]
```diff
@@ -3,6 +3,7 @@ targetCompatibility = 1.6
 
 dependencies {
     compile project(':core-api')
+    testCompile project(':core-api').sourceSets.test.output
```
Comment (mdodsworth) [inline core-httpclient-impl/build.gradle]: necessary to pull in `ExhaustiveTest` category (even though it's not explicitly used by the core-httpclient-impl module test

**Comment by mdodsworth** [comment]
ping

**Comment by mdodsworth** [comment]
@vraja2 sure. I've rebased against that branch and all is well. Once v2-support is merged in, I'll send out the revised pull request

**Comment by wangjoshuah** [comment]
@vraja2 are you following up on this?

**Comment by mikeng13** [comment]
CLosing this old branch out.

**Comment by aliabbasrizvi** [comment]
Assinging this to

**Comment by thomaszurkan-optimizely** [comment]
merged with https://github.com/optimizely/java-sdk/pull/236

**Pull request closed without merging**